File: ocaml/src/GRL_1_A.ml
(*+ import include/template/template.ml +*)
(*+ import include/graph/dijkstra.ml +*)

let () =
  let v = read_int () in
  let e = read_int () in
  let r = read_int () in
  let g = Array.make v [] in
  for i = 0 to e - 1 do
    let s = read_int () in
    let t = read_int () in
    let d = read_int () in    
    add_edge g s t d
  done;
  let d = dijkstra int_info g r in
  for i = 0 to v - 1 do
    let res =
      if d.(i) = int_info.inf
      then "INF"
      else string_of_int d.(i) in
    print_endline res
  done

File: cpp/src/GRL_1_A.cpp
#include <cstdio>

#include "../include/graph/weighted_graph.hpp"
#include "../include/graph/dijkstra.cpp"

int main() {
  int V, E, r, s, t, d;
  scanf("%d%d%d", &V ,&E, &r);
  WeightedGraph<int> g(V);
  while (E--) {
    scanf("%d%d%d", &s ,&t, &d);
    g.add(s, t, d);
  }
  auto res = dijkstra(g, r);
  for (int i = 0; i < V; ++i) {
    if (res[i] == inf<int>) puts("INF");
    else printf("%d\n", res[i]);
  }
  return 0;
}
